### PR TITLE
Packaging updates

### PR DESCRIPTION
**Pull request by drdavella**
## Overview
*Make some incremental updates to the packaging and metadata*

⚠️ This PR will probably cause conflicts with lots of in-progress work. I'm happy to hold off on this PR for a bit and rebase when we're ready, but I also don't want to wait too long. ⚠️ 

## Description
* This is an incremental step to make some updates to the structure and metadata of the Python package itself
* There are [good motivations](https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure) for using src-based layouts, so I just went ahead and made that change
* One consequence of doing this is the package needs to be installed prior to testing and linting. This is a Good Thing™ since it means we are testing code in the way it is actually installed/used
  * I also needed to update the way we look for coverage
* Also it seems like the community is standardizing on the use of `pyproject.toml` for packaging metadata
  * This could enable us to explore alternative build tools like [poetry](https://python-poetry.org/) if we were so inclined
* I made an update so that our semgrep files are packaged using `package-data` and can be accessed using `importlib.resources`
* Made some other small cosmetic updates to pipeline steps

## Future Work
* I'm creating an epic to track this, but there are some more things we want to do:
  * Add more package metadata
  * Use script entry point for `codemodder` command instead of `python -m codemodder`
  * Build the package using `python -m build` and test the result using `twine`
  * Use tag-based versioning and releases
  * Release to PyPi

### CODE REVIEW

File: pyproject.toml
Comment (clavedeluna) [inline pyproject.toml:8]: should be 3.8 no?

File: pyproject.toml
Comment (clavedeluna) [inline pyproject.toml:11]: I assume these are the dependencies that are currently in requirements/codemodder.txt? Or what is the relationship between this and that file? Can we delete the file?

File: tests/transformations/test_clean_imports.py
Comment (clavedeluna) [inline tests/transformations/test_clean_imports.py]: how did this change come about?

File: src/codemodder/semgrep.py
Comment (clavedeluna) [inline src/codemodder/semgrep.py:12]: this is magic ✨

**Review by clavedeluna — COMMENTED** [review]
Once I can go ahead and merge the other 2 pending PRs, I think we can hold off and let you merge this one.

File: pyproject.toml
Comment (drdavella) [reply pyproject.toml:8]: Sorry I meant to leave a comment here but personally I'd like to just go with 3.9 unless there are strong objections. It narrows our support window, especially with 3.12 on the horizon, and I don't see a compelling reason to go back that far.

File: pyproject.toml
Comment (drdavella) [reply pyproject.toml:11]: Yes we can delete it 👍

File: tests/transformations/test_clean_imports.py
Comment (drdavella) [reply tests/transformations/test_clean_imports.py]: Sorry, another place I meant to leave a comment. Short answer is: I have no idea. I don't think I touched any code that would have any affect on this. The new version looks reasonable/correct to me, but I'm just not sure.

Comment (clavedeluna) [reply tests/transformations/test_clean_imports.py]: sure! I Don't think we have committed to 3.8 in any official capacity or in documentation so that's good with me.

**Review by andrecsilva — APPROVED** [review]
Sounds good. Just remember to update CONTRIBUTING.md to reflect these changes.

**Pull request merged**
